rajawali/src/main/java/c/org/rajawali3d/engine/EngineImpl.java

```diff
@@ -220,105 +244,104 @@ public void onRenderSurfaceCreated(EGLConfig config, GL10 gl, int width, int hei
 
     @Override
     public void onRenderSurfaceDestroyed(SurfaceTexture surface) {
-        stopRendering();
+        stopFrames();
     }
 
     @Override
     public void onRenderSurfaceSizeChanged(GL10 gl, int width, int height) {
-        defaultViewportWidth = width;
-        defaultViewportHeight = height;
+        surfaceWidth = width;
+        surfaceHeight = height;
 
-        synchronized (renderables) {
-            for (Renderable renderable : renderables) {
-                renderable.onRenderSurfaceSizeChanged();
+        synchronized (surfaceCallbacks) {
+            for (SurfaceCallback surfaceListener : surfaceCallbacks) {
+                surfaceListener.onSurfaceSizeChanged(width, height);
             }
         }
 
-        startRendering();
+        startFrames();
     }
 
+    /**
+     * TODO Override this method in a stereo renderer to call renderViews() for each Eye
+     *
+     * @param gl {@link GL10} for rendering.
+     */
     @Override
     public void onRenderFrame(GL10 gl) {
-        synchronized (nextRenderableLock) {
-            // Check if we need to switch the scene, and if so, do it.
-            if (nextRenderable != null) {
-                switchRenderable(nextRenderable);
-                nextRenderable = null;
-            }
-        }
 
         final long currentTime = System.nanoTime();
-        final long elapsedRenderTime = currentTime - renderStartTime;
-        final double deltaTime = (currentTime - lastRender) / 1e9;
-        lastRender = currentTime;
+        final double deltaTime = (currentTime - lastFrameTime) / 1e9;
+        lastFrameTime = currentTime;
 
-        try {
-            onRender(elapsedRenderTime, deltaTime);
-        } catch (InterruptedException e) {
-            Log.e(TAG, "Frame skipped due to thread interruption.");
-        }
+        startFrame(deltaTime);
+
+        renderViews();
+
+        endFrame(deltaTime);
 
         ++frameCount;
         if (frameCount % 50 == 0) {
             long now = System.nanoTime();
-            double elapsedS = (now - startTime) / 1.0e9;
+            double elapsedS = (now - fpsStartTime) / 1.0e9;
             double msPerFrame = (1000 * elapsedS / frameCount);
             lastMeasuredFPS = 1000 / msPerFrame;
 
             frameCount = 0;
-            startTime = now;
+            fpsStartTime = now;
 
             if (fpsUpdateListener != null)
                 fpsUpdateListener.onFPSUpdate(lastMeasuredFPS); // Update the FPS listener
         }
     }
 
-    @Override
-    public void onOffsetsChanged(float xOffset, float yOffset, float xOffsetStep, float yOffsetStep, int xPixelOffset, int yPixelOffset) {
-
-    }
+    protected void startFrame(double deltaTime) {
```

Comment:
OK on FloatRange, will also incorporate IntRange if we decide to replace elapsed time.